### ISSUE TITLE
Add persistent implementation of `dataset::MutableDataset`

### ISSUE THREAD

**Issue by pchampin** [pchampin/sophia_rs#22]
In addition to the [`dataset::inmem`](https://docs.rs/sophia/0.3.0/sophia/dataset/inmem/index.html) module, it would be nice to have a disk-based persistent implementation of [`dataset::MutableDataset`](https://docs.rs/sophia/0.3.0/sophia/dataset/trait.MutableDataset.html).

**Comment by pchampin** [comment]
One way to do it could be to use [RockDB](https://rocksdb.org/). We could even try and use the same layout as used by [Oxigraph](https://github.com/Tpt/oxigraph), making it possible to share the same storage across both crates. @Tpt, what do you think? Is that layout documented somewhere?

**Comment by Tpt** [comment]
That would be great to make Sophia works with Oxigraph storage! The Oxigraph RockDB layout is not stable yet, I am currently tweaking it to make it a bit more compact and allow efficient range queries. I hope to have time finishing a 0.1 Oxigraph release with a stable RocksDB layout in late December or (more realistically) January.

The basic storage approach should not change: I store in RocksDB keys rotations of quads of [`EncodedTerm`](https://github.com/Tpt/oxigraph/blob/e0bbe29dc2377fc1277d4c232c9c26c90bd13551/lib/src/store/numeric_encoder.rs#L93) and a string store. RocksDB prefix searches are then used to solve triple patterns. The string store is used as an inverse hash lookup, the strings being hashed inside of `EncodedTerm`. Hashing strings is very useful for heavy SPARQL query evaluations with a lot of joins, it might not be the best approach for just storing and doing simple triple pattern evaluation.

But, I'm not sure that reimplementing Oxigraph storage in Sophia is the best way to do it. I fear than very quickly you might want also to be able to run SPARQL queries on top of it in Sophia, completely duplicating the work already done in Oxigraph. A better way to go would probably to make Oxigraph usable with Sophia.

**Comment by pchampin** [comment]
> A better way to go would probably to make Oxigraph usable with Sophia.

Yes, implementing Sophia's traits *above* Oxigraph is also a way to go, and probably the fastest one. I would be concerned, though, that converting from Oxigraph's model to Sophia's would induce some overhead, hence my initial proposal... But definitely worth a try, anyway.

**Comment by pchampin** [comment]
Not published on crates.io yet, but an adapter for Oxigraph is now available at https://github.com/pchampin/sophia_oxigraph.

**Comment by pchampin** [comment]
postponing this to a later milestone (if any).
My hope is that eventually, [R2C2](https://github.com/w3c-cg/r2c2) will provide interoperability between Sophia and [Oxigraph](https://github.com/oxigraph/oxigraph).
